cpp/tests/copying/shift_tests.cpp

```diff
@@ -96,10 +96,22 @@ TYPED_TEST(ShiftTest, OneColumn)
 {
   using T = TypeParam;
 
-  auto input = fixed_width_column_wrapper<T>{lowest<T>, T(1), T(2), T(3), T(4), T(5), highest<T>};
-  auto expected = fixed_width_column_wrapper<T>{T(7), T(7), lowest<T>, T(1), T(2), T(3), T(4)};
-
-  auto fill   = make_scalar<T>(T(7));
+  auto input    = fixed_width_column_wrapper<T>{lowest<T>,
+                                             cudf::test::make_type_param_scalar<T>(1),
```

Comment:
What's with all the `make_type_param_scalar`? I thought the whole point of adding the converting type parameter to `fixed_width_column_wrapper` was to avoid this.

Reply:
there are  typed tests that are run with all the fixed width types. within those tests, an instance of the type is created using an integer. hence, this method creates the appropriate type using the types constructor, if the integer value cannot be converted to the target type. 

the `fixed_width_column_wrapper` changes was to create a column of fixed width type with integer values. `make_type_param_scalar` creates a single fixed width type with an integer, as timestamps can no longer be created an integer value.

here is one example:

https://github.com/rapidsai/cudf/blob/353e39a74052fee11710c0715531de553ba5009d/cpp/tests/copying/scatter_tests.cpp#L809

Reply:
I understand that not all types can be created from integers. I thought that was the point of the changes you'd already made to `fixed_width_column_wrapper` by adding the second explicit template argument, e.g., I would think the same thing could be accomplished like this:
```
auto input = fixed_width_column_wrapper<T, int32_t>{ lowest<T>, 1, 2, 3, 4, 5, highest<T> };
```

I thought that was the point of this code: https://github.com/rapidsai/cudf/blob/f836f8ff903f472d6819876067dd0490bddf94e8/cpp/tests/utilities/column_wrapper.hpp#L139-L157

Reply:
>> `auto input = fixed_width_column_wrapper<T, int32_t>{ lowest<T>, 1, 2, 3, 4, 5, highest<T> };`

the issue here is that a `fixed_width_column_wrapper` is attempted to be created with mixed types in the initializer list - `lowest<T>` and `highest<T>` returns a `T`, where `T` can be a `timestamp`, `duration` etc. while the other items (`1..5`) are integers. hence, before an attempt is even made to convert items of type `SourceElementT` to `ElementTo` in `fixed_width_column_wrapper` using the `fixed_width_type_converter`, this invocation has to resolve first to a constructor in `fixed_width_column_wrapper` that takes in `Ts` and `ints`, `T` being *any* fixed width type, and there aren't any.

```
template <typename ElementTo, typename SourceElementT = ElementTo>
class fixed_width_column_wrapper : public detail::column_wrapper {
```

further, this isn't the only place that warranted this change in the first place. there are a number of places, *notably* in typed tests, where one or more of the following is done. here are some examples:

- create a `scalar<T>` (numeric/duration/timestamp etc.) with `T` created out of an integer
- comparing scalar `value()` (returning an object of type `T`) with an object of type `T` created using an integer value
- creating a fixed width type with an integer, passing it to a template function, where the value is compared to another `T` - creating `vector<T>` of size `n` with an object of type `T` created using an integer value etc.

Reply:
I understand there are situations where you need to construct a _single_ element of any type `T` from an integer. However, that's not what is going on here. You're constructing _many_ elements. This is extremely smelly code:

```
  auto expected = fixed_width_column_wrapper<T>{cudf::test::make_type_param_scalar<T>(7),
                                                cudf::test::make_type_param_scalar<T>(7),
                                                lowest<T>,
                                                cudf::test::make_type_param_scalar<T>(1),
                                                cudf::test::make_type_param_scalar<T>(2),
                                                cudf::test::make_type_param_scalar<T>(3),
                                                cudf::test::make_type_param_scalar<T>(4)};
```

We need a different solution.

Reply:
this was the original code:

`auto expected = fixed_width_column_wrapper<T>{T(7), T(7), lowest<T>, T(1), T(2), T(3), T(4)};`

are you expecting the callers to use something such as this:

`auto expected = fixed_width_column_wrapper<T, int32_t>{7, 7, lowest<T>, 1, 2, 3, 4};`

>> You're constructing many elements. 

so, is the objection here that the `fixed_width_column_wrapper` is constructed with many `T`s with some `T`s constructed using `make_type_param_scalar`? the original test code also constructed many elements. can you please clarify?

Reply:
The objection is to the fact that the new solution requires 20x more characters than the old solution just to intialize a column of 7 elements. The code is vastly more complex looking than it needs to be.

Reply:
This PR has dragged on long enough. We'll just live with this solution for now.

Reply:
>> The objection is to the fact that the new solution requires 20x more characters than the old solution just to intialize a column of 7 elements.

would something like this work?

`auto input = fixed_width_column_wrapper<T, int64_t>{lowest<T, int64_t>(), 1L, 2L, 3L, 4L, 5L, highest<T, int64_t>()};`